nexus-repository-p2/src/test/java/org/sonatype/nexus/repository/p2/internal/util/JarParserTest.java

```diff
@@ -53,34 +56,46 @@ public void getVersionFromJarInputStream() throws Exception {
     when(tempBlob.get()).thenReturn(getClass().getResourceAsStream(JAR_NAME));
     JarInputStream jis = new JarInputStream(tempBlob.get());
 
-    assertThat(getAttributesFromJarFile(jis).getComponentVersion(), is(equalTo("1.2.100.v20170912-1859")));
+    P2Attributes attributesFromJarFile = underTest.getAttributesFromFeatureXML(jis).get();
+    assertThat(attributesFromJarFile.getComponentVersion(), is(equalTo("1.2.100.v20170912-1859")));
+  }
+
+  @Test
+  public void getVersionFromManifestJarInputStream() throws Exception {
+    when(tempBlob.get()).thenReturn(getClass().getResourceAsStream(JAR_NAME_WITH_MANIFEST));
+    JarInputStream jis = new JarInputStream(tempBlob.get());
+
+    P2Attributes attributesFromJarFile = getAttributesFromJarFile(jis);
+    assertThat(attributesFromJarFile.getComponentVersion(), is(equalTo("1.7.5")));
   }
 
   @Test
   public void getVersionFromSourceJar() throws Exception {
     when(tempBlob.get()).thenReturn(getClass().getResourceAsStream(JAR_SOURCES_NAME));
     JarInputStream jis = new JarInputStream(tempBlob.get());
 
-    assertThat(getAttributesFromJarFile(jis).getComponentVersion(), is(equalTo("4.7.0.v20170712-1432")));
+    P2Attributes attributesFromJarFile = getAttributesFromJarFile(jis);
+    assertThat(attributesFromJarFile.getComponentVersion(), is(equalTo("4.7.0.v20170712-1432")));
   }
 
   @Test
   public void getExceptionFromJarInputStream() throws Exception {
```

Comment:
Invalid test name - we don't check Exception

Reply:
There are few methods which throw exceptions, so we need keep it in signature. I can change to throws IOException, InvalidMetadataException. Is it pass for you?

Reply:
Done